// @ai-sdk/replicate@0.1.10/dist/index.mjs
// src/replicate-provider.ts
import { NoSuchModelError } from "@ai-sdk/provider";
import { loadApiKey } from "@ai-sdk/provider-utils";

// src/replicate-image-model.ts
import {
  combineHeaders,
  createBinaryResponseHandler,
  createJsonResponseHandler,
  getFromApi,
  postJsonToApi,
  resolve
} from "@ai-sdk/provider-utils";
import { z as z2 } from "zod";

// src/replicate-error.ts
import { createJsonErrorResponseHandler } from "@ai-sdk/provider-utils";
import { z } from "zod";
var replicateErrorSchema = z.object({
  detail: z.string().optional(),
  error: z.string().optional()
});
var replicateFailedResponseHandler = createJsonErrorResponseHandler({
  errorSchema: replicateErrorSchema,
  errorToMessage: (error) => {
    var _a, _b;
    return (_b = (_a = error.detail) != null ? _a : error.error) != null ? _b : "Unknown Replicate error";
  }
});

// src/replicate-image-model.ts
var ReplicateImageModel = class {
  constructor(modelId, settings, config) {
    this.modelId = modelId;
    this.settings = settings;
    this.config = config;
    this.specificationVersion = "v1";
  }
  get provider() {
    return this.config.provider;
  }
  get maxImagesPerCall() {
    var _a;
    return (_a = this.settings.maxImagesPerCall) != null ? _a : 1;
  }
  async doGenerate({
    prompt,
    n,
    aspectRatio,
    size,
    seed,
    providerOptions,
    headers,
    abortSignal
  }) {
    var _a, _b, _c, _d;
    const warnings = [];
    const [modelId, version] = this.modelId.split(":");
    const currentDate = (_c = (_b = (_a = this.config._internal) == null ? void 0 : _a.currentDate) == null ? void 0 : _b.call(_a)) != null ? _c : /* @__PURE__ */ new Date();
    const {
      value: { output },
      responseHeaders
    } = await postJsonToApi({
      url: (
        // different endpoints for versioned vs unversioned models:
        version != null ? `${this.config.baseURL}/predictions` : `${this.config.baseURL}/models/${modelId}/predictions`
      ),
      headers: combineHeaders(await resolve(this.config.headers), headers, {
        prefer: "wait"
      }),
      body: {
        input: {
          prompt,
          aspect_ratio: aspectRatio,
          size,
          seed,
          num_outputs: n,
          ...(_d = providerOptions.replicate) != null ? _d : {}
        },
        // for versioned models, include the version in the body:
        ...version != null ? { version } : {}
      },
      successfulResponseHandler: createJsonResponseHandler(
        replicateImageResponseSchema
      ),
      failedResponseHandler: replicateFailedResponseHandler,
      abortSignal,
      fetch: this.config.fetch
    });
    const outputArray = Array.isArray(output) ? output : [output];
    const images = await Promise.all(
      outputArray.map(async (url) => {
        const { value: image } = await getFromApi({
          url,
          successfulResponseHandler: createBinaryResponseHandler(),
          failedResponseHandler: replicateFailedResponseHandler,
          abortSignal,
          fetch: this.config.fetch
        });
        return image;
      })
    );
    return {
      images,
      warnings,
      response: {
        timestamp: currentDate,
        modelId: this.modelId,
        headers: responseHeaders
      }
    };
  }
};
var replicateImageResponseSchema = z2.object({
  output: z2.union([z2.array(z2.string()), z2.string()])
});

// src/replicate-provider.ts
function createReplicate(options = {}) {
  const createImageModel = (modelId, settings) => {
    var _a;
    return new ReplicateImageModel(modelId, settings != null ? settings : {}, {
      provider: "replicate",
      baseURL: (_a = options.baseURL) != null ? _a : "https://api.replicate.com/v1",
      headers: {
        Authorization: `Bearer ${loadApiKey({
          apiKey: options.apiToken,
          environmentVariableName: "REPLICATE_API_TOKEN",
          description: "Replicate"
        })}`,
        ...options.headers
      },
      fetch: options.fetch
    });
  };
  return {
    image: createImageModel,
    imageModel: createImageModel,
    languageModel: () => {
      throw new NoSuchModelError({
        modelId: "languageModel",
        modelType: "languageModel"
      });
    },
    textEmbeddingModel: () => {
      throw new NoSuchModelError({
        modelId: "textEmbeddingModel",
        modelType: "textEmbeddingModel"
      });
    }
  };
}
var replicate = createReplicate();
export {
  createReplicate,
  replicate
};
